magma/array.py

```diff
@@ -438,5 +460,57 @@ def is_mixed(cls):
         return cls.T.is_mixed()
 
 
-
 ArrayType = Array
+
+
+class CoreIRCommonLibMuxN(Generator2):
+    def __init__(self, N: int, width: int):
+        self.name = f"coreir_commonlib_mux{N}x{width}"
+        MuxInT = Product.from_fields("anon", dict(data=Array[N, Array[width, Bit]],
+                                                  sel=Array[clog2(N), Bit]))
+        self.io = IO(I=In(MuxInT), O=Out(Array[width, Bit]))
+        self.renamed_ports = coreir_port_mapping
+        self.coreir_name = "muxn"
+        self.coreir_lib = "commonlib"
+        self.coreir_genargs = {"width": width, "N": N}
+        self.primitive = True
+        self.stateful = False
+
+        def simulate(self, value_store, state_store):
+            sel = BitVector[clog2(N)](value_store.get_value(self.I.sel))
+            out = BitVector[width](value_store.get_value(self.I.data[int(sel)]))
+            value_store.set_value(self.O, out)
+
+        self.simulate = simulate
+
+
+class Mux(Generator2):
+    def __init__(self, height: int, T: Type):
+        if issubclass(T, MagmaProtocol):
+            T = T._to_magma_()
+        T_str = str(T).replace("(", "").replace(")", "")\
+                      .replace(",", "_").replace("=", "_")\
+                      .replace("[", "").replace("]", "").replace(" ", "")
+        self.name = f"Mux{height}x{T_str}"
+        # TODO: Type must be hashable so we can cache.
+        N = T.flat_length()
+
+        io_dict = {}
+        for i in range(height):
+            io_dict[f"I{i}"] = In(T)
+        if height == 2:
+            io_dict["S"] = In(Bit)
+        else:
+            io_dict["S"] = In(Array[clog2(height), Bit])
+        io_dict["O"] = Out(T)
+
+        self.io = IO(**io_dict)
+        mux = CoreIRCommonLibMuxN(height, N)()
+        data = [Array[N, Bit](getattr(self.io, f"I{i}").flatten())
+                for i in range(height)]
+        mux.I.data @= Array[height, Array[N, Bit]](data)
+        if height == 2:
+            mux.I.sel[0] @= self.io.S
+        else:
+            mux.I.sel @= self.io.S
+        self.io.O @= T.unflatten(mux.O)
```

Comment:
This seems like an odd place to have this, perhaps we can have a primitive directory and have a Mux file imported here?

Reply:
Yes, ideally would have liked to have `mux.py` or a separate file, but unfortunately `Mux` class depends on `Array` and `Array` depends on `Mux` class through `dynamic_mux_select` function. Fixes the dependency issue to have them in the same file, but I agree it's bad organization.

The fundamental issue is that there are operators we want on the types that involve "higher level" things, e.g. the `__getitem__` on `Array` invoking mux. The two alternatives I see are, (a) allowing the operators to inject functions - I think this was how things were done before ala monkey patching: `Array.__getitem__ = fn` in `mux.py`; and (b) only *declaring* those functions at operator time so like

```python
def dynamic_mux_select(self, ...):
    Select = DeclareCircuit('Select', ...)
    select = Select()
    select.I @= ...
    return select.O
```

Reply:
Linking is probably the right long term solution (declare interface, get implementation later).  Monkey patching is probably a reasonable short-term solution since it's low code overhead and I think I'd prefer that over having the mux operator inside the array file (since mux should work with Bit/Tuple as well, so it seems odd to have it in array)

Reply:
Moved to monkey patching, seems reasonable for now.